doc/man/fedfs-null.8
.\"@(#)fedfs-null.8"
.\"
.\" @file doc/man/fedfs-null.8
.\" @brief man page for fedfs-null client command
.\"

.\"
.\" Copyright 2011 Oracle.  All rights reserved.
.\"
.\" This file is part of fedfs-utils.
.\"
.\" fedfs-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2.0 as
.\" published by the Free Software Foundation.
.\"
.\" fedfs-utils is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 2.0 for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" version 2.0 along with fedfs-utils.  If not, see:
.\"
.\"	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
.\"
.TH FEDFS-NULL 8 "@publication-date@"
.SH NAME
fedfs-null \- send a FEDFS_NULL ADMIN protocol request
.SH SYNOPSIS
.B fedfs-null
.RB [ \-?d ]
.RB [ \-n
.IR nettype ]
.RB [ \-h
.IR hostname ]
.SH INTRODUCTION
RFC 5716 introduces the Federated File System (FedFS, for short).
FedFS is an extensible standardized mechanism
by which system administrators construct
a coherent namespace across multiple file servers using
.IR "file system referrals" .
For further details, see
.BR fedfs (7).
.P
FedFS-enabled file servers allow remote administrative access via an
authenticated RPC protocol known as the
.IR "FedFS ADMIN protocol" .
Using this protocol, FedFS administrators manage
FedFS junctions and NSDB connection parameter information
on remote FedFS-enabled file servers.
.SH DESCRIPTION
The
.BR fedfs-null (8)
command is part of a collection of low-level single-use programs
that is intended for testing the FedFS ADMIN protocol or for use in scripts.
It sends a single FEDFS_NULL request to a remote
FedFS ADMIN protocol service.
.P
The FEDFS_NULL request performs a simple ping operation that determines
if there is an operational FedFS ADMIN service on the remote server.
.SH OPTIONS
.IP "\fB\-d, \-\-debug"
Enables debugging messages during operation.
.IP "\fB\-?, \-\-help"
Displays
.BR fedfs-null (8)
version information and a usage message on
.IR stderr .
.IP "\fB\-h, \-\-hostname=\fIhostname\fP"
Specifies the hostname of a remote FedFS ADMIN service.
If this option is not specified, the default value is
.BR localhost .
.IP "\fB\-n, \-\-nettype=\fInettype\fP"
Specifies the transport to use when contacting the remote FedFS ADMIN service.
Typically the
.I nettype
is one of
.B tcp
or
.BR udp .
If this option is not specified, the default value is
.BR netpath .
See
.BR rpc (3t)
for details.
.SH EXIT CODES
The remote administration service returns a value that reflects the
success of the requested operation.
.TP
.B FEDFS_OK
The operation succeeded.
.TP
.B FEDFS_ERR_NOTSUPP
The remote server does not support the specified procedure.
.SH EXAMPLES
Suppose you are the FedFS administrator of the
.I example.net
FedFS domain and that you want to know if the FedFS ADMIN service
on the file server
.IR fs.example.net
is operational.  Use:
.RS
.sp
$ fedfs-null -h fs.example.net
.br
Call completed successfully
.RE
.SH SECURITY
RPCSEC GSSAPI authentication has not yet been implemented for this command.
.SH "SEE ALSO"
.BR fedfs (7),
.BR rpc.fedfsd (8),
.BR rpc (3t)
.sp
RFC 5716 for FedFS requirements and overview
.SH COLOPHON
This page is part of the fedfs-utils package.
A description of the project and information about reporting bugs
can be found at
.IR http://wiki.linux-nfs.org/wiki/index.php/FedFsUtilsProject .
.SH "AUTHOR"
Chuck Lever <chuck.lever@oracle.com>


src/fedfsc/fedfs-null.c
/**
 * @file src/fedfsc/fedfs-null.c
 * @brief Send a NULL RPC to a FedFS ADMIN server
 */

/*
 * Copyright 2010 Oracle.  All rights reserved.
 *
 * This file is part of fedfs-utils.
 *
 * fedfs-utils is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2.0 as
 * published by the Free Software Foundation.
 *
 * fedfs-utils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License version 2.0 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2.0 along with fedfs-utils.  If not, see:
 *
 *	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>
#include <locale.h>

#include <rpc/clnt.h>

#include "fedfs.h"
#include "fedfs_admin.h"
#include "xlog.h"
#include "gpl-boiler.h"

/**
 * Default RPC request timeout
 */
static struct timeval fedfs_null_timeout = { 25, 0 };

/**
 * Short form command line options
 */
static const char fedfs_null_opts[] = "?dh:n:";

/**
 * Long form command line options
 */
static const struct option fedfs_null_longopts[] = {
	{ "debug", 0, NULL, 'd', },
	{ "help", 0, NULL, '?', },
	{ "hostname", 1, NULL, 'h', },
	{ "nettype", 1, NULL, 'n', },
	{ NULL, 0, NULL, 0, },
};

static void
fedfs_null_usage(const char *progname)
{
	fprintf(stderr, "\n%s version " VERSION "\n", progname);
	fprintf(stderr, "Usage: %s [-d] [-n nettype] [-h hostname]\n\n",
			progname);

	fprintf(stderr, "\t-?, --help           Print this help\n");
	fprintf(stderr, "\t-d, --debug          Enable debug messages\n");
	fprintf(stderr, "\t-n, --nettype        RPC transport (default: 'netpath')\n");
	fprintf(stderr, "\t-h, --hostname       ADMIN server hostname (default: 'localhost')\n");
	fflush(stderr);

	fprintf(stderr, "%s", fedfs_gpl_boilerplate);

	exit((int)FEDFS_ERR_INVAL);
}

static FedFsStatus
fedfs_null_call(const char *hostname, const char *nettype)
{
	FedFsStatus exit_status;
	enum clnt_stat status;
	CLIENT *client;
	char result;

	client = clnt_create(hostname, FEDFS_PROG, FEDFS_V1, nettype);
	if (client == NULL) {
		clnt_pcreateerror("Failed to create FEDFS client");
		return -1;
	}

	exit_status = FEDFS_OK;
	memset((char *)&result, 0, sizeof(result));
	status = clnt_call(client, FEDFS_NULL,
				(xdrproc_t)xdr_void, (caddr_t)NULL,
				(xdrproc_t)xdr_void, (caddr_t)&result,
				fedfs_null_timeout);
	if (status != RPC_SUCCESS) {
		clnt_perror(client, "FEDFS_NULL call failed");
		exit_status = FEDFS_ERR_SVRFAULT;
	} else
		printf("Call completed successfully\n");

	(void)clnt_destroy(client);
	return exit_status;
}

int
main(int argc, char **argv)
{
	char *progname, *hostname, *nettype;
	unsigned int seconds;
	FedFsStatus status;
	int arg;

	(void)setlocale(LC_ALL, "");
	(void)umask(S_IRWXO);

	/* Set the basename */
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	/* For the libraries */
	xlog_stderr(1);
	xlog_syslog(0);
	xlog_open(progname);

	hostname = "localhost";
	nettype = "netpath";
	while ((arg = getopt_long(argc, argv, fedfs_null_opts, fedfs_null_longopts, NULL)) != -1) {
		switch (arg) {
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 'h':
			hostname = optarg;
			break;
		case 'n':
			nettype = optarg;
			break;
		default:
			fprintf(stderr, "Invalid command line argument: %c\n", (char)arg);
		case '?':
			fedfs_null_usage(progname);
		}
	}
	if (optind != argc)
		fedfs_null_usage(progname);

	for (seconds = FEDFS_DELAY_MIN_SECS;; seconds = fedfs_delay(seconds)) {
		status = fedfs_null_call(hostname, nettype);
		if (status != FEDFS_ERR_DELAY)
			break;

		xlog(D_GENERAL, "Delaying %u seconds...", seconds);
		if (sleep(seconds) != 0)
			break;
	}
	return (int)status;
}
